File: flappy_goud/Pipe.cs
// Pipe.cs

using System;
using CsBindgen;

public class Pipe
{
    private readonly GoudGame game;
    public uint topSpriteId;
    public uint bottomSpriteId;
    public float X { get; private set; }
    public float GapY { get; private set; }

    public uint TextureId { get; private set; }

    // Img Dimensions: 52 × 320
    public Pipe(GoudGame game)
    {
        this.game = game;
        X = GameConstants.ScreenWidth;

        // Set the gap to a random vertical position on the screen
        GapY = new Random().Next(GameConstants.PipeGap, (int)GameConstants.ScreenHeight - GameConstants.PipeGap);
        TextureId = game.CreateTexture("assets/sprites/pipe-green.png");

        // Position the top pipe above the screen by its height
        topSpriteId = game.AddSprite(new SpriteCreateDto
        {
            x = X,
            y = GapY - GameConstants.PipeGap - 320, // Adjusted to position the top pipe correctly
            rotation = 180,
            texture_id = TextureId
        });


        bottomSpriteId = game.AddSprite(new SpriteCreateDto
        {
            x = X,
            y = GapY + GameConstants.PipeGap, // Adjusted to position the bottom pipe correctly
            texture_id = TextureId
        });

    }

    public void Update(float deltaTime)
    {
        X -= GameConstants.PipeSpeed * deltaTime * GameConstants.TargetFPS;

        // Update top and bottom pipe positions
        game.UpdateSprite(new SpriteUpdateDto { id = topSpriteId, x = X, y = GapY - GameConstants.PipeGap - 320, rotation = 180, texture_id = TextureId });
        game.UpdateSprite(new SpriteUpdateDto { id = bottomSpriteId, x = X, y = GapY + GameConstants.PipeGap, texture_id = TextureId });
    }

    public bool IsOffScreen()
    {
        return X + GameConstants.PipeWidth < 0;
    }



    public bool IsPassed(float birdX)
    {
        return birdX > X + GameConstants.PipeWidth;
    }

    public bool Remove()
    {
        game.RemoveSprite(topSpriteId);
        game.RemoveSprite(bottomSpriteId);
        return true;
    }
}

File: examples/kotlin/flappy_bird/src/main/kotlin/Pipe.kt
import com.goudengine.core.GoudGame
import com.goudengine.types.Color
import kotlin.random.Random

/**
 * A pair of top/bottom pipes scrolling leftward.
 */
class Pipe private constructor(
    private val game: GoudGame,
    var x: Float,
    private val gapY: Float,
    private val pipeTexture: Long,
) {
    var scored = false

    fun update(dt: Float) {
        x = Movement.scrollLeft(x, GameConstants.PIPE_SPEED, dt)
    }

    fun render(game: GoudGame) {
        val pw = GameConstants.PIPE_WIDTH.toFloat()
        val gap = GameConstants.PIPE_GAP.toFloat()
        val topH = gapY
        val bottomY = gapY + gap
        val bottomH = GameConstants.SCREEN_HEIGHT.toFloat() - bottomY

        // Top pipe
        game.drawSprite(pipeTexture, x, 0f, pw, topH, 0f, Color.white())
        // Bottom pipe
        game.drawSprite(pipeTexture, x, bottomY, pw, bottomH, 0f, Color.white())
    }

    fun isOffScreen(): Boolean = x + GameConstants.PIPE_WIDTH < 0

    fun collidesWith(birdX: Float, birdY: Float): Boolean {
        val bw = GameConstants.BIRD_WIDTH.toFloat()
        val bh = GameConstants.BIRD_HEIGHT.toFloat()
        val pw = GameConstants.PIPE_WIDTH.toFloat()
        val gap = GameConstants.PIPE_GAP.toFloat()

        if (birdX + bw < x || birdX > x + pw) return false
        if (birdY < gapY || birdY + bh > gapY + gap) return true
        return false
    }

    companion object {
        fun spawn(game: GoudGame): Pipe {
            val minY = 50f
            val maxY = (GameConstants.SCREEN_HEIGHT - GameConstants.BASE_HEIGHT - GameConstants.PIPE_GAP - 50).toFloat()
            val gapY = Random.nextFloat() * (maxY - minY) + minY
            val texture = game.loadTexture("assets/sprites/pipe-green.png")
            return Pipe(game, GameConstants.SCREEN_WIDTH.toFloat(), gapY, texture)
        }
    }
}
